src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/EnumShouldNotHaveDuplicatedValuesTests.cs

```diff
@@ -0,0 +1,411 @@
+{
+    Value1 = 1,
+    Value2 = 1,
+}
+
+public enum MyUShortEnum : ushort
+{
+    Value1 = 1,
+    Value2 = 1,
+}
+
+public enum MyIntEnum : int
+{
+    Value1 = 1,
+    Value2 = 1,
+}
+
+public enum MyUIntEnum : uint
+{
+    Value1 = 1,
+    Value2 = 1,
+}
+
+public enum MyLongEnum : long
+{
+    Value1 = 1,
+    Value2 = 1,
+}
+
+public enum MyULongEnum : ulong
+{
+    Value1 = 1,
+    Value2 = 1,
+}",
+                GetCSharpResultAt(5, 5),
+                GetCSharpResultAt(11, 5),
+                GetCSharpResultAt(17, 5),
+                GetCSharpResultAt(23, 5),
+                GetCSharpResultAt(29, 5),
+                GetCSharpResultAt(35, 5),
+                GetCSharpResultAt(41, 5),
+                GetCSharpResultAt(47, 5));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyByteEnum As Byte
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MySByteEnum As SByte
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MyShortEnum As Short
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MyUShortEnum As UShort
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MyIntEnum As Integer
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MyUIntEnum As UInteger
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MyLongEnum As Long
+    Value1 = 1
+    Value2 = 1
+End Enum
+
+Public Enum MyULongEnum As ULong
+    Value1 = 1
+    Value2 = 1
+End Enum
+",
+                GetBasicResultAt(4, 5),
+                GetBasicResultAt(9, 5),
+                GetBasicResultAt(14, 5),
+                GetBasicResultAt(19, 5),
+                GetBasicResultAt(24, 5),
+                GetBasicResultAt(29, 5),
+                GetBasicResultAt(34, 5),
+                GetBasicResultAt(39, 5));
+        }
+
+        [Fact]
+        public async Task EnumExplicitBitwiseShiftDuplication_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    None = 0x0,
+    Value1 = 0x1,
+    Value2 = 0x1,
+    Value3 = 0x4,
+}",
+                GetCSharpResultAt(6, 5));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    None = &H0
+    Value1 = &H1
+    Value2 = &H1
+    Value3 = &H4
+End Enum",
+                GetBasicResultAt(5, 5));
+        }
+
+        [Fact]
+        public async Task EnumExplicitDuplicationNotConsecutive_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2 = 2,
+    Value3 = 1,
+}",
+                GetCSharpResultAt(6, 5));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    None = &H0
+    Value1 = &H1
+    Value2 = &H1
+    Value3 = &H4
+End Enum",
+                GetBasicResultAt(5, 5));
+        }
+
+        [Fact]
+        public async Task EnumImplicitDuplication_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2,
```

Comment:
This case is no longer handled as there is no field initializer here.

I am actually wondering if it might be useful to have a change on Roslyn to have a callback for the default initialization of a field with some flag (maybe implicit) to detect user written from auto-generated ones. WDYT?

The only solution I found to handle this case isn't right IMO as it consists of a list initialized with the enum fields, for which I'll remove each initialized field in the operation block. In the end action, I will add to the dictionary all uninitialized fields with their values.

Reply:
ping @mavasani

Reply:
> I am actually wondering if it might be useful to have a change on Roslyn to have a callback for the default initialization of a field with some flag (maybe implicit) to detect user written from auto-generated ones.

That is a good idea, can you open a Roslyn issue and tag me on it?

> The only solution I found to handle this case isn't right IMO as it consists of a list initialized with the enum fields, for which I'll remove each initialized field in the operation block. In the end action, I will add to the dictionary all uninitialized fields with their values.

I don't have a better idea either. I am fine with your proposed approach as a workaround until your previous suggestion about callback with implicit default initialization is implemented in Roslyn.

Reply:
https://github.com/dotnet/roslyn/issues/40811